lucene/queries/src/java/org/apache/lucene/queries/function/valuesource/QueryValueSource.java

```diff
@@ -156,42 +135,38 @@ public float floatVal(int doc) {
   @Override
   public boolean exists(int doc) {
     try {
-      if (doc < lastDocRequested) {
-        if (noMatches) return false;
+      if (noMatches) return false;
+      if (scorer == null) {
         scorer = weight.scorer(readerContext);
-        scorerDoc = -1;
-        if (scorer==null) {
+        if (scorer == null) {
           noMatches = true;
           return false;
         }
-        it = scorer.iterator();
+        tpi = scorer.twoPhaseIterator();
+        disi = tpi==null ? scorer.iterator() : tpi.approximation();
+        thisDocMatches = false;
       }
+      assert doc >= lastDocRequested : "values requested out of order; last=" + lastDocRequested + ", requested=" + doc;
```

Comment:
Hmm; I wasn't sure about this but the more I look, it appears that the ValueSource API (here) also has a forward-only requirement.  A few years ago this was not the case.  So we can assume this and throw an exception.  See JoinDocFreqValueSource for an example.  I think this check you added here should be way up front at this method, thus clearly showing as a pre-condition, not down here.

Reply:
I also didn't think so until your last review, but I found the following on the docstring for `ValueSource.getValues`:

> The values must be consumed in a forward docID manner, and you must call this method again to iterate through the values again.